src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -206,7 +211,7 @@ async ValueTask<AsyncLazy<Solution>> GetLazySolutionAndIncrementRefCountAsync()
                         // We're the first call that is asking about this checksum.  Create a lazy to compute it with a
                         // refcount of 1 (for 'us').
                         tuple = (refCount: 1, AsyncLazy.Create(
-                            c => ComputeSolutionAsync(assetProvider, solutionChecksum, workspaceVersion, fromPrimaryBranch, c), cacheResult: true));
+                            c => ComputeSolutionAsync(assetProvider, solutionChecksum, currentSolution, c), cacheResult: true));
```

Comment:
this was the core issue.  teh asynclazy computation would capture workspaceVersion/fromPrimaryBranch.  If fromPRimaryBranch was false, this computation would not actually update the workspace.  But other concurrent calls that did want to update it would then just see that htis async lazy was in flight and do nothing.  

Now, all the concurrent calls will share this computation.  But then the ones that want to update the workspace will try to do that no matter what at https://github.com/dotnet/roslyn/pull/60749/files#diff-fedbb1bbc0cee7c7ea4dbd51b8b27023e351ec9a062fb4623fdb8063b5517d30R179.